experimental/framework/fluid-llm/src/shared-tree-diff/sharedTreeBranchManager.ts

```diff
@@ -0,0 +1,261 @@
+import { type TreeArrayNode } from "@fluidframework/tree";
+import type * as z from "zod";
+
+import {
+	sharedTreeDiff,
+	type Difference,
+	type DifferenceChange,
+	type DifferenceCreate,
+	type DifferenceMove,
+	type DifferenceRemove,
+	type ObjectPath,
+} from "./sharedTreeDiff.js";
+import { isTreeMapNode, isTreeArrayNode, sharedTreeTraverse } from "./utils.js";
+
+/**
+ * Manages determining the differences between two branches of a SharedTree represented as an actual tree node or a plain javascript object
+ * and applies said differences to the original SharedTree branch.
+ */
+export class SharedTreeBranchManager {
+	private readonly objectSchema?: z.Schema;
+	private readonly nodeIdAttributeName?: string;
+
+	public constructor(params?: { objectSchema?: z.Schema; nodeIdAttributeName?: string }) {
+		this.objectSchema = params?.objectSchema;
+		this.nodeIdAttributeName = params?.nodeIdAttributeName;
+	}
+
+	public compare(
+		obj: Record<string, unknown> | TreeArrayNode,
+		newObj: Record<string, unknown> | unknown[],
+	): Difference[] {
+		// By validating that the incoming object matches the schema, we can confirm that any property
+		// deletions/updates/additions are valid.
+		if (this.objectSchema !== undefined) {
+			const res = this.objectSchema.safeParse(newObj);
+			if (res.success === false) {
+				throw new TypeError("Invalid data");
```

Comment:
Would probably be good to log a bit more here. I think `zod` gives you back more specific notes about what parts of the data were incompatible with the schema, right?

Reply:
Yes zod gives very rich information about what exactly is incorrect. I've got a helper function which I'm planning to bring in from one of my own projects that will concatenate them to produce a nice string